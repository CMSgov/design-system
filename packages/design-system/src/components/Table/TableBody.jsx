import PropTypes from 'prop-types';
import React from 'react';

export const TableBody = ({ children, _stackable, ...tableBodyProps }) => {
  const renderChildren = () => {
    return React.Children.map(children, (child) => {
      // Extend props before rendering.
      // TODO: Use React Context when all products are on React v16.8 or higher
      if (child && child.props) {
        return React.cloneElement(child, {
          _stackable: _stackable,
        });
      }
      return child;
    });
  };

  /* eslint-disable jsx-a11y/no-redundant-roles */
  return (
    <tbody role="rowgroup" {...tableBodyProps}>
      {_stackable ? renderChildren() : children}
    </tbody>
  );
  /* eslint-enable jsx-a11y/no-redundant-roles */
};

TableBody.propTypes = {
  /**
   * The table body contents, usually `TableRow`.
   */
  children: PropTypes.node,
  /**
   * @hide-prop This gets set from the parent `Table` component
   */
  _stackable: PropTypes.bool,
};

export default TableBody;
