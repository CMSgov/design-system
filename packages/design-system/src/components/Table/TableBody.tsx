import React from 'react';
export interface TableBodyProps {
  /**
   * The table body contents, usually `TableRow`.
   */
  children?: React.ReactNode;
}

type OmitProps = 'children';

/**
 * `TableBody` renders the `<tbody>` element and will typically contain `TableRow` elements to define table data.
 */
export const TableBody: React.FC<
  Omit<React.ComponentPropsWithoutRef<'tbody'>, OmitProps> & TableBodyProps
> = ({ children, ...tableBodyProps }: TableBodyProps) => {
  return (
    <tbody role="rowgroup" {...tableBodyProps}>
      {children}
    </tbody>
  );
};

export default TableBody;
