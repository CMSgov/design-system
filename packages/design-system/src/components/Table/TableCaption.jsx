import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';

export const TableCaption = ({
  children,
  className,
  _id,
  _scrollActive,
  _scrollableNotice,
  ...tableCaptionProps
}) => {
  const classes = classNames('ds-c-table__caption', className);
  return (
    <caption className={classes} id={_id} {...tableCaptionProps}>
      {children}
      {_scrollActive && _scrollableNotice}
    </caption>
  );
};

// Set component name to make child.type.displayName available to other components (eg. Table)
TableCaption.displayName = 'TableCaption';

TableCaption.propTypes = {
  /**
   * The table caption contents.
   */
  children: PropTypes.node,
  /**
   * Additional classes to be added to the caption element.
   */
  className: PropTypes.string,
  /**
   * @hide-prop This gets passed from the parent `Table` component when the table `scrollable` prop is set.
   */
  _id: PropTypes.string,
  /**
   * @hide-prop This gets passed from the parent `Table` component when the table `scrollable` prop is set.
   */
  _scrollActive: PropTypes.bool,
  /**
   * @hide-prop This gets passed from the parent `Table` component when the table `scrollable` prop is set.
   */
  _scrollableNotice: PropTypes.node,
};

export default TableCaption;
