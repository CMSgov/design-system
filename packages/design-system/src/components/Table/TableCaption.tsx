import type * as React from 'react';
import classNames from 'classnames';

export interface TableCaptionProps {
  /**
   * The table caption contents.
   */
  children?: React.ReactNode;
  /**
   * Additional classes to be added to the caption element.
   */
  className?: string;
}

export const TableCaption = ({ children, className, ...tableCaptionProps }: TableCaptionProps) => {
  const classes = classNames('ds-c-table__caption', className);
  return (
    <caption className={classes} {...tableCaptionProps}>
      {children}
    </caption>
  );
};

// Set component name to make child.type.displayName available to other components (eg. Table)
TableCaption.displayName = 'TableCaption';

export default TableCaption;
