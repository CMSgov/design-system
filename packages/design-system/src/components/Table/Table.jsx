import Alert from '../Alert/Alert';
import PropTypes from 'prop-types';
import React from 'react';
import TableCaption from './TableCaption';
import classNames from 'classnames';
import get from 'lodash/get';
import uniqueId from 'lodash.uniqueid';

// TODO: Revert out of this 'PR update to use lifecycle methods'
// (https://github.com/CMSgov/design-system/pull/777)
// when hc.gov child ds and the product apps are on react v16.8

function debounce(fn, ms) {
  let timer;
  return () => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      timer = null;
      fn.apply(this, arguments);
    }, ms);
  };
}

/**
 * Determine if a React component is a TableCaption
 * @param {React.Node} child - a React component
 * @return {Boolean} Is this a TableCaption component?
 */
function isTableCaption(child) {
  const componentName = get(child, 'type.displayName') || get(child, 'type.name');

  // Check child.type first and as a fallback, check child.type.displayName follow by child.type.name
  return child && (child.type === TableCaption || componentName === 'TableCaption');
}

export class Table extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      scrollActive: false,
    };
    this.captionID = uniqueId('caption-');
    this.container = 0;
    this.debounceHandleResize = debounce(this.handleResize.bind(this), 500);

    if (process.env.NODE_ENV !== 'production') {
      if (
        props.scrollable &&
        Array.isArray(props.children) &&
        !props.children.some((child) => isTableCaption(child))
      ) {
        console.warn(
          'The children prop in `Table` must include `TableCaption` component for scrollable tables.'
        );
      }
      if (props.dense) {
        console.warn(
          `[Deprecated]: Please remove the 'dense' prop in <Table>, use 'compact' instead. This prop has been renamed and will be removed in a future release.`
        );
      }
    }
  }

  componentDidMount() {
    if (this.props.scrollable) {
      window.addEventListener('resize', this.debounceHandleResize);
      this.handleResize();
    }
  }

  componentWillUnmount() {
    if (this.props.scrollable) {
      window.removeEventListener('resize', this.debounceHandleResize);
    }
  }

  handleResize() {
    const { scrollWidth, clientWidth } = this.container;
    const scrollActive = scrollWidth > clientWidth;
    this.setState({ scrollActive });
  }

  renderChildren() {
    return React.Children.map(this.props.children, (child) => {
      if (isTableCaption(child)) {
        // Extend props on TableCaption before rendering.
        // TODO: Use React Context when all products are on React v16.8 or higher
        if (this.props.scrollable) {
          return React.cloneElement(child, {
            _id: this.captionID,
            _scrollActive: this.state.scrollActive,
            _scrollableNotice: this.props.scrollableNotice,
          });
        }
      } else if (this.props.stackable && child.props) {
        // Extend props for others before rendering.
        return React.cloneElement(child, {
          _stackable: this.props.stackable,
        });
      }
      return child;
    });
  }

  render() {
    const {
      borderless,
      className,
      compact,
      dense,
      stackable,
      stackableBreakpoint,
      striped,
      scrollable,
      scrollableNotice,
      children,
      ...tableProps
    } = this.props;

    const classes = classNames(
      'ds-c-table',
      borderless ? 'ds-c-table--borderless' : null,
      compact || dense ? 'ds-c-table--compact' : null,
      striped ? 'ds-c-table--striped' : null,
      stackable ? `ds-c-${stackableBreakpoint}-table--stacked` : null,
      className
    );

    // Makes table container focusable and displays the scrollable notice when table width exceeds viewport
    // by setting `tabIndex = 0` attribute.
    // This provides context for screen readers to the table's <caption> via aria-labelleby
    const attributeScrollable = scrollable && {
      className: 'ds-c-table__wrapper',
      role: 'region',
      'aria-labelledby': this.captionID,
      'aria-live': 'polite',
      'aria-relevant': 'additions',
      tabIndex: this.state.scrollActive ? '0' : null,
    };

    return (
      <div
        ref={(container) => {
          this.container = container;
        }}
        {...attributeScrollable}
      >
        <table className={classes} role="table" {...tableProps}>
          {this.renderChildren()}
        </table>
      </div>
    );
  }
}

Table.defaultProps = {
  scrollableNotice: (
    <Alert className="ds-u-margin-y--1 ds-u-font-weight--normal" role="status">
      <p className="ds-c-alert__text">Scroll using arrow keys to see more</p>
    </Alert>
  ),
  stackableBreakpoint: 'sm',
};

Table.propTypes = {
  /**
   * Applies the borderless variation of the table.
   */
  borderless: PropTypes.bool,
  /**
   * The table contents, usually `TableCaption`, `TableHead` and `TableBody`.
   */
  children: PropTypes.node.isRequired,
  /**
   * Additional classes to be added to the root table element.
   */
  className: PropTypes.string,
  /**
   * Applies the compact variation of the table.
   */
  compact: PropTypes.bool,
  /**
   * @hide-prop [Deprecated] Use compact instead.
   */
  dense: PropTypes.bool,
  /**
   * Applies a horizontal scrollbar and scrollable notice on `TableCaption` when the `Table`'s contents exceed the container width.
   */
  scrollable: PropTypes.bool,
  /**
   * Additional text or content to display when the horizontal scrollbar is visible to give the user notice of the scroll behavior.
   * This prop will only be used when the `Table` `scrollable` prop is set and the table width is wider than the viewport.
   */
  scrollableNotice: PropTypes.node,
  /**
   * Enables responsive styles to vertically stack rows at the specified `stackableBreakpoint`.
   * When `stackable` is set, `id` or `headers` prop is required in `TableCell`
   */
  stackable: PropTypes.bool,
  /**
   * The viewport size at which responsive vertically stacked row styles are applied. Only used when the `stackable` prop is set to `true`. [Read more on breakpoints](/guidelines/responsive/)
   */
  stackableBreakpoint: PropTypes.oneOf(['sm', 'md', 'lg']),
  /**
   * Applies the striped variation of the table.
   */
  striped: PropTypes.bool,
};

export default Table;
