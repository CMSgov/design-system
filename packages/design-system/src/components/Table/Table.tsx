import Alert from '../Alert/Alert';
import React from 'react';
import TableCaption from './TableCaption';
import TableContext from './TableContext';
import classNames from 'classnames';
import get from 'lodash/get';
import uniqueId from 'lodash/uniqueId';
import debounce from '../utilities/debounce';

export type TableStackableBreakpoint = 'sm' | 'md' | 'lg';

export interface TableProps {
  /**
   * Applies the borderless variation of the table.
   */
  borderless?: boolean;
  /**
   * The table contents, usually `TableCaption`, `TableHead` and `TableBody`.
   */
  children: React.ReactNode;
  /**
   * Additional classes to be added to the root table element.
   */
  className?: string;
  /**
   * Applies the compact variation of the table.
   */
  compact?: boolean;
  /**
   * A unique ID prefix for all the table caption elements.
   */
  id?: string;
  /**
   * Applies a horizontal scrollbar and scrollable notice on `TableCaption` when the `Table`'s contents exceed the container width.
   */
  scrollable?: boolean;
  /**
   * Additional text or content to display when the horizontal scrollbar is visible to give the user notice of the scroll behavior.
   * This prop will only be used when the `Table` `scrollable` prop is set and the table width is wider than the viewport.
   */
  scrollableNotice?: React.ReactNode;
  /**
   * A stackable variation of the table.
   * When `stackable` is set, `id` or `headers` prop is required in `Table`
   */
  stackable?: boolean;
  /**
   * Applies responsive styles to vertically stacked rows at different viewport sizes.
   */
  stackableBreakpoint?: TableStackableBreakpoint;
  /**
   * A striped variation of the table.
   */
  striped?: boolean;
  /**
   * Disables the warning message on development console when a responsive stackable table cell does not contain an `id` or `headers`.
   * It's recommended that accessibility with screen readers is tested to ensure the stacked table meets the requirement.
   */
  warningDisabled?: boolean;
}

type OmitProps = 'children' | 'className';

/**
 * Determine if a React component is a TableCaption
 * @param {React.Node} child - a React component
 * @return {Boolean} Is this a TableCaption component?
 */
function isTableCaption(child: React.ReactElement): boolean {
  const componentName = get(child, 'type.displayName') || get(child, 'type.name');

  // Check child.type first and as a fallback, check child.type.displayName follow by child.type.name
  return child && (child.type === TableCaption || componentName === 'TableCaption');
}

/**
 * `Table` is a container component that contains `TableCaption`, `TableHead`
 * and `TableBody` as children, as well as `TableRow` and `TableCell` for the
 * table content. These components mostly follow ordinary HTML table semantics,
 * but also include some additional responsive features including horizontal
 * scrolling and vertically stacked rows.
 *
 * For information about how and when to use this component,
 * [refer to its full documentation page](https://design.cms.gov/components/table/).
 */
export class Table extends React.Component<
  Omit<React.ComponentPropsWithoutRef<'table'>, OmitProps> & TableProps,
  any
> {
  static defaultProps = {
    scrollableNotice: (
      <Alert className="ds-c-table__scroll-alert" role="status">
        <p className="ds-c-alert__text">Scroll using arrow keys to see more</p>
      </Alert>
    ),
    stackableBreakpoint: 'sm',
  };

  constructor(props: TableProps) {
    super(props);
    this.state = {
      scrollActive: false,
    };
    this.captionId = props.id ?? uniqueId('table-caption--');
    this.container = 0;
    this.debounceHandleResize = debounce(this.handleResize.bind(this), 500);

    if (process.env.NODE_ENV !== 'production') {
      if (
        props.scrollable &&
        Array.isArray(props.children) &&
        !props.children.some((child: React.ReactElement) => isTableCaption(child))
      ) {
        console.warn(
          'The children prop in `Table` must include `TableCaption` component for scrollable tables.'
        );
      }
    }
  }

  componentDidMount(): void {
    if (this.props.scrollable) {
      window.addEventListener('resize', this.debounceHandleResize);
      this.handleResize();
    }
  }

  componentWillUnmount(): void {
    if (this.props.scrollable) {
      window.removeEventListener('resize', this.debounceHandleResize);
    }
  }

  captionId: string;
  container: any;
  debounceHandleResize: (...args: any[]) => any;

  handleResize(): void {
    const { scrollWidth, clientWidth } = this.container;
    const scrollActive = scrollWidth > clientWidth;
    this.setState({ scrollActive });
  }

  renderChildren(): React.ReactNode[] {
    return React.Children.map(this.props.children, (child: React.ReactElement) => {
      if (isTableCaption(child)) {
        // Extend props on TableCaption before rendering.
        if (this.props.scrollable) {
          return React.cloneElement(child, {
            _id: this.captionId,
            _scrollActive: this.state.scrollActive,
            _scrollableNotice: this.props.scrollableNotice,
          });
        }
      }
      return child;
    });
  }

  render() {
    const {
      borderless,
      className,
      compact,
      stackable,
      stackableBreakpoint,
      striped,
      scrollable,
      scrollableNotice,
      warningDisabled,
      children,
      id,
      ...tableProps
    } = this.props;

    const classes = classNames(
      'ds-c-table',
      borderless ? 'ds-c-table--borderless' : null,
      compact ? 'ds-c-table--compact' : null,
      striped ? 'ds-c-table--striped' : null,
      stackable ? `ds-c-${stackableBreakpoint}-table--stacked` : null,
      className
    );

    /**
     * Makes table container focusable and displays the scrollable notice when table width exceeds viewport
     * by setting `tabIndex = 0` attribute.
     * This provides context for screen readers to the table's <caption> via aria-labelleby
     */
    const attributeScrollable = scrollable && {
      className: 'ds-c-table__wrapper',
      role: 'region',
      'aria-labelledby': this.captionId,
      tabIndex: this.state.scrollActive ? 0 : null,
    };
    const contextValue = { stackable: !!stackable, warningDisabled: !!warningDisabled };

    return scrollable ? (
      <div
        ref={(container) => {
          this.container = container;
        }}
        aria-live="polite"
        aria-relevant="additions"
        {...attributeScrollable}
      >
        <TableContext.Provider value={contextValue}>
          <table className={classes} role="table" {...tableProps}>
            {this.renderChildren()}
          </table>
        </TableContext.Provider>
      </div>
    ) : (
      <TableContext.Provider value={contextValue}>
        <table className={classes} role="table" {...tableProps}>
          {this.renderChildren()}
        </table>
      </TableContext.Provider>
    );
  }
}

export default Table;
