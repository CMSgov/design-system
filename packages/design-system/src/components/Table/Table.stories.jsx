/* eslint-disable react/no-array-index-key */
/* eslint-disable react/no-multi-comp */
import React from 'react';

import Table from './Table';
import TableCaption from './TableCaption';
import TableHead from './TableHead';
import TableRow from './TableRow';
import TableCell from './TableCell';
import TableBody from './TableBody';

export default {
  title: 'Components/Table',
  component: Table,
  argTypes: {
    dense: {
      table: {
        disable: true,
      },
    },
  },
};

const tableTemplateData = {
  headings: [
    { displayName: 'Document title', propName: 'documentTitle' },
    { displayName: 'Description', propName: 'description' },
    { displayName: 'Links', propName: 'links' },
    { displayName: 'Year', propName: 'year' },
  ],
  data: [
    {
      documentTitle: 'Declaration of Independence',
      description:
        'Statement adopted by the Continental Congress declaring independence from the British Empire.',
      links: (
        <a href="# ">
          {' '}
          https://billofrightsinstitute.org/founding-documents/declaration-of-independence/
        </a>
      ),
      year: '1776',
    },
    {
      documentTitle: 'Bill of Rights',
      description:
        'The first ten amendments of the U.S. Constitution guaranteeing rights and freedoms.',
      links: <a href="# ">https://billofrightsinstitute.org/founding-documents/bill-of-rights/</a>,
      year: '1791',
    },
  ],
};

const multiHeaderTableData = {
  headings: [
    { displayName: 'Name', propName: 'name', bodyComponent: 'th' },
    { displayName: 'Street', propName: 'street' },
    { displayName: 'ZIP code', propName: 'zipcode' },
    { displayName: 'Employer', propName: 'employer' },
    { displayName: 'Industry', propName: 'industry' },
  ],
  data: [
    {
      name: 'John Doe',
      street: '123 Braavos Ave.',
      zipcode: '20005',
      employer: 'Acme Co.',
      industry: 'Healthcare',
    },
    {
      name: 'John Doe',
      street: "456 King's Landing",
      zipcode: '20005',
      employer: 'Acme Co.',
      industry: 'Healthcare',
    },
  ],
};

const Template = (args) => (
  <Table {...args}>
    <TableCaption>Table</TableCaption>
    <TableHead>
      <TableRow>
        {tableTemplateData.headings.map(({ displayName }) => (
          <TableCell key={displayName}>{displayName}</TableCell>
        ))}
      </TableRow>
    </TableHead>
    <TableBody>
      {tableTemplateData.data.map((dataItem, index) => (
        <TableRow key={index}>
          {tableTemplateData.headings.map((heading) => (
            <TableCell
              key={`${heading.displayName}-${dataItem[heading.propName]}`}
              stackedTitle={heading.displayName}
            >
              {dataItem[heading.propName]}
            </TableCell>
          ))}
        </TableRow>
      ))}
    </TableBody>
  </Table>
);

export const ScrollableTable = Template.bind({});

export const StackableTable = Template.bind({});
StackableTable.args = {
  stackable: true,
  stackableBreakpoint: 'md',
};

export const MultiHeaderTable = () => (
  <Table>
    <TableCaption>Household members</TableCaption>
    <col />
    <colgroup span="2" />
    <colgroup span="2" />
    <TableHead>
      <TableRow>
        <TableCell component="td" rowSpan="1" />
        <TableCell component="th" colSpan="2" scope="colgroup">
          Address
        </TableCell>
        <TableCell colSpan="2" scope="colgroup">
          Employment
        </TableCell>
      </TableRow>
      <TableRow>
        {multiHeaderTableData.headings.map(({ displayName }) => (
          <TableCell component="th" scope="col" key={displayName}>
            {displayName}
          </TableCell>
        ))}
      </TableRow>
    </TableHead>
    <TableBody>
      {multiHeaderTableData.data.map((dataItem, index) => (
        <TableRow key={index}>
          {multiHeaderTableData.headings.map((heading) => (
            <TableCell
              key={`${heading.displayName}-${dataItem[heading.displayName]}`}
              component={heading.bodyComponent || 'td'}
            >
              {dataItem[heading.propName]}
            </TableCell>
          ))}
        </TableRow>
      ))}
    </TableBody>
  </Table>
);
