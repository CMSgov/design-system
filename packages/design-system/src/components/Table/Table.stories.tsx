import React from 'react';
import Table from './Table';
import TableCaption from './TableCaption';
import TableHead from './TableHead';
import TableRow from './TableRow';
import TableCell from './TableCell';
import TableBody from './TableBody';
import type { Meta, StoryObj } from '@storybook/react';

const meta: Meta<typeof Table> = {
  title: 'Components/Table',
  component: Table as any,
  subcomponents: {
    TableCaption,
    TableHead,
    TableRow,
    TableCell,
    TableBody,
  },
};
export default meta;

type Story = StoryObj<typeof Table>;

const tableTemplateData = {
  headings: [
    { displayName: 'Document title', propName: 'documentTitle' },
    { displayName: 'Description', propName: 'description' },
    { displayName: 'Links', propName: 'links' },
    { displayName: 'Year', propName: 'year' },
  ],
  data: [
    {
      documentTitle: 'Declaration of Independence',
      description:
        'Statement adopted by the Continental Congress declaring independence from the British Empire.',
      links: (
        <a href="# ">
          {' '}
          https://billofrightsinstitute.org/founding-documents/declaration-of-independence/
        </a>
      ),
      year: '1776',
    },
    {
      documentTitle: 'Bill of Rights',
      description:
        'The first ten amendments of the U.S. Constitution guaranteeing rights and freedoms.',
      links: <a href="# ">https://billofrightsinstitute.org/founding-documents/bill-of-rights/</a>,
      year: '1791',
    },
  ],
};

const multiHeaderTableData = {
  headings: [
    { displayName: 'Name', propName: 'name', bodyComponent: 'th' },
    { displayName: 'Street', propName: 'street' },
    { displayName: 'ZIP code', propName: 'zipcode' },
    { displayName: 'Employer', propName: 'employer' },
    { displayName: 'Industry', propName: 'industry' },
  ],
  data: [
    {
      name: 'John Doe',
      street: '123 Braavos Ave.',
      zipcode: '20005',
      employer: 'Acme Co.',
      industry: 'Healthcare',
    },
    {
      name: 'John Doe',
      street: "456 King's Landing",
      zipcode: '20005',
      employer: 'Acme Co.',
      industry: 'Healthcare',
    },
  ],
};

const Default: Story = {
  render: function Component(args) {
    return (
      <Table {...args}>
        <TableCaption>Table</TableCaption>
        <TableHead>
          <TableRow>
            {tableTemplateData.headings.map(({ displayName, propName }) => (
              <TableCell key={displayName} id={propName}>
                {displayName}
              </TableCell>
            ))}
          </TableRow>
        </TableHead>
        <TableBody>
          {tableTemplateData.data.map((dataItem, index) => (
            <TableRow key={index}>
              {tableTemplateData.headings.map((heading) => (
                <TableCell
                  key={`${heading.displayName}-${dataItem[heading.propName]}`}
                  headers={heading.propName}
                  stackedTitle={heading.displayName}
                >
                  {dataItem[heading.propName]}
                </TableCell>
              ))}
            </TableRow>
          ))}
        </TableBody>
      </Table>
    );
  },
};

export const ScrollableTable: Story = {
  ...Default,
  args: {
    scrollable: true,
  },
};

export const StackableTable: Story = {
  ...Default,
  args: {
    stackable: true,
    stackableBreakpoint: 'md',
  },
};

export const MultiHeaderTable: Story = {
  render: function Component(args) {
    return (
      <Table>
        <TableCaption>Household members</TableCaption>
        <col />
        <colgroup span={2} />
        <colgroup span={2} />
        <TableHead>
          <TableRow>
            <TableCell component="td" rowSpan={1} />
            <TableCell component="th" colSpan={2} scope="colgroup">
              Address
            </TableCell>
            <TableCell colSpan={2} scope="colgroup">
              Employment
            </TableCell>
          </TableRow>
          <TableRow>
            {multiHeaderTableData.headings.map(({ displayName }) => (
              <TableCell component="th" scope="col" key={displayName}>
                {displayName}
              </TableCell>
            ))}
          </TableRow>
        </TableHead>
        <TableBody>
          {multiHeaderTableData.data.map((dataItem, index) => (
            <TableRow key={index}>
              {multiHeaderTableData.headings.map((heading) => (
                <TableCell
                  key={`${heading.displayName}-${dataItem[heading.displayName]}`}
                  component={heading.bodyComponent || ('td' as any)}
                >
                  {dataItem[heading.propName]}
                </TableCell>
              ))}
            </TableRow>
          ))}
        </TableBody>
      </Table>
    );
  },
};
