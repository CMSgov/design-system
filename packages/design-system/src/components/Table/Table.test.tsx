import React from 'react';
import { render, screen } from '@testing-library/react';

import Table from './Table';
import TableCaption from './TableCaption';

const makeTable = (customProps = {}) => {
  const children = <TableCaption>A great caption</TableCaption>;
  render(
    <Table id="static-id" {...customProps}>
      {children}
    </Table>
  );
};

describe('Table', function () {
  it('sets role="table"', () => {
    makeTable();
    expect(screen.getAllByRole('table')).toHaveLength(1);
  });

  it('renders table with class ds-c-table', () => {
    makeTable();
    expect(screen.getByRole('table')).toHaveClass('ds-c-table');
  });

  it('applies additional classNames to root table', () => {
    makeTable({ className: 'foo-table' });
    expect(screen.getByRole('table')).toHaveClass('foo-table');
  });

  it('applies borderless classes', () => {
    makeTable({ borderless: true });
    expect(screen.getByRole('table')).toHaveClass('ds-c-table--borderless');
  });

  it('applies compact table classes', () => {
    makeTable({ compact: true });
    expect(screen.getByRole('table')).toHaveClass('ds-c-table--compact');
  });

  it('applies zebra stripe classes', () => {
    makeTable({ striped: true });
    expect(screen.getByRole('table')).toHaveClass('ds-c-table--striped');
  });

  it('accepts custom id', () => {
    makeTable({ scrollable: true });
    const el = screen.getByRole('region');
    expect(el).toMatchSnapshot();
  });

  it('applies responsive stacked table', () => {
    makeTable({ stackable: true, stackableBreakpoint: 'lg' });
    const table = screen.getByRole('table');
    expect(table).toHaveClass('ds-c-lg-table--stacked');
    expect(table).toMatchSnapshot();
  });

  it('applies scrollable prop to table', () => {
    makeTable({ scrollable: true });
    const scrollableRegion = screen.getByRole('region');

    expect(scrollableRegion).toHaveClass('ds-c-table__wrapper');
    expect(scrollableRegion).toHaveAttribute('aria-live', 'polite');
    expect(scrollableRegion).toHaveAttribute('aria-relevant', 'additions');
    expect(scrollableRegion).toMatchSnapshot();
  });

  it('scroll table aria-labelledby matches caption id', () => {
    makeTable({ scrollable: true });
    const region = screen.getByRole('region').attributes['id'];
    const caption = screen.getByText('A great caption').attributes['aria-labelledby'];

    expect(region).toEqual(caption);
  });
});
