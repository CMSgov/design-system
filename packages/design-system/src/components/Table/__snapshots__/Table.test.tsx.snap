// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table accepts custom id 1`] = `
<div
  aria-labelledby="static-id"
  aria-live="polite"
  aria-relevant="additions"
  class="ds-c-table__wrapper"
  role="region"
>
  <table
    class="ds-c-table"
    role="table"
  >
    <caption
      class="ds-c-table__caption"
      id="static-id"
    >
      A great caption
    </caption>
  </table>
</div>
`;

exports[`Table applies responsive stacked table 1`] = `
<table
  class="ds-c-table ds-c-lg-table--stacked"
  role="table"
>
  <caption
    class="ds-c-table__caption"
  >
    A great caption
  </caption>
</table>
`;

exports[`Table applies scrollable prop to table 1`] = `
<div
  aria-labelledby="static-id"
  aria-live="polite"
  aria-relevant="additions"
  class="ds-c-table__wrapper"
  role="region"
>
  <table
    class="ds-c-table"
    role="table"
  >
    <caption
      class="ds-c-table__caption"
      id="static-id"
    >
      A great caption
    </caption>
  </table>
</div>
`;
